Add support for reading arcinfo adf tin layers #162

vcloarec opened the pull request:
First Draft, the mesh seems to be loaded. Need tests to verify if the geometry is ok.

saberraz commented:
@vcloarec nice. 
I tried it with the attached mesh.

```
mdalinfo tnz.adf
Mesh loaded: OK
  Driver: ESRI_TIN
  Vertex count: 13
  Face count: 19
  Projection: undefined
Datasets loaded: OK
  Groups count: 1
  Terrain Elevation
```

But when I try to open it in QGIS, I get core dump.
[top.zip](https://github.com/lutraconsulting/MDAL/files/3715702/top.zip)

PS: we can probably use the attached TIN for test and credit http://lab.usgin.org

PeterPetrik commented on mdal/frmts/mdal_esri_tin.cpp:
```diff
@@ -0,0 +1,179 @@
+#include "mdal_esri_tin.hpp"
+
+//from https://stackoverflow.com/questions/41012414/convert-double-value-from-little-endian-to-big-endian
+template <typename T>
```
I think we have a function for this in https://github.com/lutraconsulting/MDAL/blob/master/mdal/frmts/mdal_selafin.cpp#L145

maybe we can just move that function to MDAL utils and use it here too?

vcloarec replied:
I have added new function in MDAL utils, and I have changed selafin driver and esri tin driver.
Tell me what do you think about ?
But I am not sure that it is a good idea because the number of bytes read in the file depend now of the size of type. Is there a risk that this size of type is different from platform to other ? If yes, it is maybe not a good idea ...

PeterPetrik commented on mdal/frmts/mdal_esri_tin.cpp:
```diff
@@ -0,0 +1,179 @@
+    if ( ! faceIn.is_open() )
+        return false;
+
+    //check if the file sizes of the nodes are coherent ?
```
i think it does not need to be... what you can check is first line of some file if they have some standard header/first line

vcloarec replied:
There isn't more information that the position of the points and the indexes of the face. The unused first bytes (64 bytes for tnxy.adf, and 16 bytes for tnz.adf) are used by ArcGIS to store information about 4 virtual points created by ArcGIS for triangulation/generation (https://en.wikipedia.org/wiki/Esri_TIN, see Superpoint). Those points seem to be necessary for ArcGIS but not for MDAL.

So, for me, the only information to know if a file can be read is :
- if it can be opened
- if the size of the tns.adf file (Z value file) is coherent with the size of of tnxy.adf file.
I am not sure of the relevance of checking the size file coherence.

PeterPetrik replied:
lets keep it as it is, just remove commented out code please

PeterPetrik commented on mdal/frmts/mdal_esri_tin.cpp:
```diff
@@ -0,0 +1,179 @@
+
+void MDAL::DriverEsriTin::populateVertices( const std::string &uri, MDAL::Vertices &vertices )
+{
+    //firt, read the (x,y) coordinates
```
typo: first

PeterPetrik commented on mdal/frmts/mdal_esri_tin.cpp:
```diff
@@ -0,0 +1,179 @@
+    if ( ! inXY.is_open() )
+        throw MDAL_Status::Err_FileNotFound;
+
+    inXY.ignore( 64 ); //ignore the 64 first bytes
```
this is probably something to function canRead(). aren't those 64bytes some common string?

vcloarec replied:
No, see the other comment

PeterPetrik commented on mdal/frmts/mdal_esri_tin.cpp:
```diff
@@ -0,0 +1,179 @@
+        throw MDAL_Status::Err_FileNotFound;
+
+    inXY.ignore( 64 ); //ignore the 64 first bytes
+    //bool readingOk=(!inXY.eof());
```
remove commented out code

PeterPetrik commented on mdal/frmts/mdal_esri_tin.cpp:
```diff
@@ -0,0 +1,179 @@
+    {
+        Vertex vert;
+
+        if ( !readAndChangeEndianness( vert.x, inXY ) )
```
this is hard to read. I would extract the in.read to this loop and handle reading double in another routine.

vcloarec replied:
Sorry, I am not sure to understand this note.

PeterPetrik replied:
my point is that reading of the stream should be outside of the function. One function - one task... 

other question was how it behaves if you host system has big endianness too... I was refering to selafin implementation , see https://github.com/lutraconsulting/MDAL/blob/master/mdal/frmts/mdal_selafin.cpp#L145

PeterPetrik commented on mdal/frmts/mdal_esri_tin.cpp:
```diff
@@ -0,0 +1,179 @@
+    throw MDAL_Status::Err_FileNotFound;
+
+  inZ.ignore( 16 ); //ignore the 16 first bytes
+  //readingOk=(!inZ.eof());
```
remove

PeterPetrik commented on mdal/frmts/mdal_esri_tin.cpp:
```diff
@@ -0,0 +1,179 @@
+    if ( f.size() < 3 ) //that's mean the face is not complete
+      throw MDAL_Status::Err_UnknownFormat;
+
+    //adjust the indexes and exclude face with "superpoints
```
missing "

what is superpoint? can you add description here?

vcloarec replied:
Description is in the top of `mdal/frmts/mdal_esri_tin.hpp` file but I will try to be more explicit.

PeterPetrik replied:
thanks

PeterPetrik commented on mdal/frmts/mdal_esri_tin.hpp:
```diff
@@ -0,0 +1,79 @@
+   *
+   *
+   *
+   * other bytess: X,Y : for each value 8 bits : double with big endianness
```
typo

PeterPetrik commented on mdal/frmts/mdal_esri_tin.hpp:
```diff
@@ -0,0 +1,79 @@
+   *
+   *
+  */
+
```
remove extra space/newlines all over the file

PeterPetrik commented on mdal/mdal_utils.cpp:
```diff
@@ -487,7 +488,7 @@ void MDAL::addBedElevationDatasetGroup( MDAL::Mesh *mesh, const Vertices &vertic
                                           mesh,
                                           mesh->uri(),
-                                          "Bed Elevation"
+                                          "Terrain Elevation"
```
Please keep Bed elevation. We are using the term for this for 10 years...

PeterPetrik commented on mdal/mdal_utils.cpp:
```diff
@@ -505,6 +506,12 @@ void MDAL::addBedElevationDatasetGroup( MDAL::Mesh *mesh, const Vertices &vertic
   mesh->datasetGroups.push_back( group );
 }
 
+void MDAL::addBedElevationDatasetGroup( MDAL::Mesh *mesh, const Vertices &vertices )
```
I would just use Bed elevation, not Terrain elevation . It is standard

vcloarec replied:
For me "Bed Elevation" is specific for river bed elevation. Terrain Elevation could be more generic. As TIN are note only for river, that's why I purpose to keep Bed Elevation for format specific to hydraulic modeling, and maybe we can purpose another name for other format not specific to river.

PeterPetrik replied:
whatabout keeping the Bed elevation and just change the name in the TIN driver...?

PeterPetrik commented on tests/CMakeLists.txt:
```diff
@@ -72,7 +72,7 @@ SET(TESTS
     test_mdal_utils.cpp
     test_selafin.cpp
-)
+    test_esri_tin.cpp)
```
missing newline before )

PeterPetrik commented on tests/test_esri_tin.cpp:
```diff
@@ -0,0 +1,38 @@
+
+  int f_count = MDAL_M_faceCount( m );
+  EXPECT_EQ( f_count, 19 );
+
```
please add test for few faces to see if the vertices are correctly assigned. Also add check for bed elevation dataset

nyalldawson commented:
Wow, great work.... That was super fast!!

vcloarec commented:
@saberraz 
The issue is may be fix, it was bad indexes for faces

vcloarec commented:
@PeterPetrik , I try to look at all your comments may be later today.

saberraz commented:
Nice, my test mesh works now!
![test](https://user-images.githubusercontent.com/2649047/66671457-efbb9700-ec53-11e9-9772-91c21e772247.png)

vcloarec commented:
> Nice, my test mesh works now!
It is not the prettier TIN I've seen :)
The non-public TIN sent by Nyall look better!

nyalldawson commented:
I've got some nice simple reference files incoming (freely licensed). Will post a link when I receive them!

vcloarec commented:
> I've got some nice simple reference files incoming (freely licensed). Will post a link when I receive them!

Is it possible to have TIN files with holes ?

nyalldawson commented:
I'll ask. Also trying to get some with tagged edges/vertices...

vcloarec commented:
Hello,
With this last commit, I think MDAL reads ESRI TIN as well as possible.  Holes haven't been tested, but with coming Nyall files, maybe it will be done.
My comments on particular points below :

**Superpoints :**
If the first bytes of xy file and z file are usually used for particular points (called "superpoints" here) that are not used by MDAL, it seems it is not always the case. The file "thul.adf" contains the indexes of the superpoints that can be no the first. The code has been adapted to take into account superpoints that are not at the 
beginning. But as we don't have a file with superpoints not at the beginning, it is not tested in such a situation.

**Mask :**
ArcGIS uses the mask to make invisible faces, those which are outside the outer boundaries, in holes or linked with superpoints. In this PR, MDAL simply used the mask file to ignore those faces.

**Name of the TIN :**
ESRI TINs are saved in a folder that the user gives when he creates it in ArcGis. All the different files in this folder always have the same names and have all the extension *.adf. As MDAL and QGIS used a filename to load a mesh, the trick is to retrieve the folder with any files with *.adf extension is provided to MDAL. So, in QGIS,  selecting any *.adf files permits to load the TIN. The bad consequence is the name of the mesh is the name of the selected file, not the name of the folder...
I have implemented a method in the ESRI TIN driver to retrieve the name of the TIN. But now, it is not used because it seems not trivial for QGIS to obtain this name...

**Endianness :**
The endianness is handled by a new function in MDAL utils. This template function permits to read the value of any type in a binary file. An option provides the possibility to change the endianness of the value if needed.
Depending on the platform, if the type size is not compatible with the same type in the file, the caller keeps the responsibility to handle with this.
For Selafin driver, I have changed the `read_double()` method to see how this method could be improved. But, as some other methods could have to be changed too, I purpose to restore the initial code of this method. The Selafin driver could be improved in another commit/PR if the new template function is pertinent.

PeterPetrik commented on mdal/frmts/mdal_esri_tin.cpp:
```diff
@@ -0,0 +1,300 @@
+  std::string crsWkt;
+  std::getline( inCRS, crsWkt );
+
+  if ( crsWkt == "{B286C06B-0879-11D2-AACA-00C04FA33C20}" )
```
I do not understand this {B28....}? 

Wouldn't it be more sensible to read projection from *.prj file? (3D_Boreholes.prj in the test case?)

nyalldawson replied:
It's a "magic sequence" used in ESRI to indicate a "not set" or unknown CRS. (It's actually the com class id of the esri UnknownCoordinateSystem class).

PeterPetrik replied:
do you know if this file contains actually WKT or it is always some magic sequence of class ids?

vcloarec replied:
I don't know why the *.prj file is here. With other TIN, there is not this file but there is always a prj.adf file. The prj.adf file contains WKT.

PeterPetrik replied:
ok no problem
, can you add comment on the line?

vcloarec replied:
Done

PeterPetrik commented:
thanks a lot, merging!

nyalldawson commented:
Awesome work -- this is very exciting!

@PeterPetrik any chance of a mdal version bump prior to QGIS 3.10.0?

nyalldawson commented:
@vcloarec here's a zipped archive of a bunch of freely-usable tin reference files

[tins.zip](https://github.com/lutraconsulting/MDAL/files/3724126/tins.zip)

It includes the original dems, the arcgis tins, and the results of exporting the tins to polygon layers (in the gdb file). 

Hope it helps verify the holes/islands cases!

vcloarec commented:
Tanks @nyalldawson , I quickly tried to load the mesh in QGIS, and all have been loaded, with holes and islands. I will add this TIN to the test folder and add tests, but later, I don't have time anymore...
I am very happy this enhancement seems to be in the next release of QGIS!

nyalldawson commented:
@vcloarec 

For completeness, the attached updates to the tins adds an example with tagged vertices. That should cover all the functionality possible in this format:

[dem (2).zip](https://github.com/lutraconsulting/MDAL/files/3737407/dem.2.zip)

PeterPetrik commented:
@nyalldawson @vcloarec any possibility you have time to add those to test suite?

vcloarec commented:
> 
> 
> @nyalldawson @vcloarec any possibility you have time to add those to test suite?

I don't have a lot of time this week for that. I will try next week.

vcloarec commented:
Hi,
I started to add the test with the TINs give by @nyalldawson. 
First test, no problem, the vertices count and the faces count read by MDAL are the same than in the gdb files.
When testing the dem with hole, there is a little difference with vertices count between MDAL and the gdb file. Two of those extra vertices are in blue in the picture below (the underlayer is the raster TIN, the source ?)
![image](https://user-images.githubusercontent.com/7416892/67541973-dd604500-f6b8-11e9-9459-0bb637c10d32.png)
 
I think, during any of the conversions, faces have gone away and the vertices stay alone... Maybe, this is due of the border of the raster TIN.

So, in the ESRI TIN files, it seems there are vertices without faces ... 
I purpose to implement a process to remove those isolated vertices, but it will slow down the loading...
What do you think about ?

nyalldawson commented:
@vcloarec I can't work out quite where in the tin you are looking -- can you show the whole extent?

nyalldawson commented:
@vcloarec disregard that, sorry!

nyalldawson commented:
@vcloarec 
Here's what I see in arcmap:

![image](https://user-images.githubusercontent.com/1829991/67545224-b860e600-f73b-11e9-8b77-4b062c32dfd8.png)

So those extra vertices are attached only to edges with a type of "Outside edge" (3). I gather you are currently stripping these away during the conversion?

vcloarec commented:
@nyalldawson, yes, outside faces (so implicitly associated edges, as MDAL doesn't store directly the edges) are stripped away.
For my information, what is the relation between the raster DEM (tif) and the TIN ? If the raster DEM is a conversion from the TIN, it is weird that some pixels are outside the boundary :
![Screenshot_20191025_082040](https://user-images.githubusercontent.com/7416892/67570949-af065800-f700-11e9-8c26-6c24aec102dd.png)


In any cases, even if there is no visual consequences to keep the isolated vertices, I think it is cleaner to strip away this vertices too when loading the mesh.
